manager/deallocator/deallocator.go

```diff
@@ -0,0 +1,244 @@
+						return err
+					}
+				}
+			}
+
+			// now we also need to look at all existing service-level resources
+			// that are marked for deletion
+			for _, resourceDeallocator := range resourceDeallocators {
+				allResources, err := resourceDeallocator.enumerator(readTx)
+
+				if err != nil {
+					log.G(ctx).WithError(err).Errorf("failed to list %vs in deallocator init", resourceDeallocator.name)
+					return err
+				}
+
+				for _, resource := range allResources {
+					if resource.isPendingDelete() {
+						if err := deallocator.maybeDeallocateResource(ctx, nil, resourceDeallocator, resource.id(), nil); err != nil {
+							return err
+						}
+					}
+				}
+			}
+
+			return nil
+		},
+		events...)
+
+	if err != nil {
+		// if we have an error here, we can't proceed any further, let the world burn
+		log.G(ctx).WithError(err).Error("failed to initialize the deallocator")
+		return err
+	}
+
+	defer func() {
+		close(deallocator.doneChan)
+		eventsChanCancel()
+	}()
+
+	// now we just need to wait for events
+	for {
+		select {
+		case event := <-eventsChan:
+			if err = deallocator.processNewEvent(ctx, event); err != nil {
+				log.G(ctx).WithError(err).Errorf("error processing deallocator event %#v", event)
+			}
+		case <-deallocator.stopChan:
+			return nil
+		}
+	}
+}
+
+func (deallocator *Deallocator) processPendingDeleteService(ctx context.Context, readTx store.ReadTx, service *api.Service) error {
+	var (
+		tasks []*api.Task
+		err   error
+	)
+
+	viewFunc := func(r store.ReadTx) {
+		tasks, err = store.FindTasks(r, store.ByServiceID(service.ID))
+	}
+
+	if readTx == nil {
+		deallocator.store.View(viewFunc)
+	} else {
+		viewFunc(readTx)
+	}
+
+	if err != nil {
+		log.G(ctx).WithError(err).Errorf("failed to retrieve the list of tasks for service %v", service.ID)
+		// if in doubt, let's proceed to clean up the service anyway
+		// better to clean up resources that shouldn't be cleaned up yet
+		// than ending up with a service and some resources lost in limbo forever
+		return deallocator.deallocateService(ctx, service)
+	} else if len(tasks) == 0 {
+		// no tasks remaining for this service, we can clean it up
+		return deallocator.deallocateService(ctx, service)
+	}
+	deallocator.services[service.ID] = serviceWithTaskCounts{service: service, taskCount: len(tasks)}
+	return nil
+}
+
+func (deallocator *Deallocator) deallocateService(ctx context.Context, service *api.Service) (err error) {
+	err = deallocator.store.Update(func(tx store.Tx) error {
+		// first, let's delete the service
+		var ignoreServiceID *string
+		if err := store.DeleteService(tx, service.ID); err != nil {
+			// all errors are just for logging here, we do a best effort at cleaning up everything we can
+			log.G(ctx).WithError(err).Errorf("failed to delete service record ID %v", service.ID)
+			ignoreServiceID = &service.ID
+		}
+
+		// then all of its service-level resources, provided no other service uses them
```

Comment:
>provided no other service uses them

How do we check that?

Reply:
That's what the `resourceDeallocator.servicesLocator` callback is for